newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,490 @@
+# Custom OTel Spans and Traces
+# ----------------------------------------------
+
+# TracerProvider: we can think of this as the agent instance.  Only one can exist
+# SpanProcessor: we can think of this as an application.  In NR, we can have multiple applications
+#   though right now, we can only do SpanProcessor and SynchronousMultiSpanProcessor
+# Tracer: we can think of this as the transaction.
+# Span: we can think of this as the trace.
+# Links functionality has now been enabled but not implemented yet.  Links are relationships
+#   between spans, but lateral in hierarchy.  In NR we only have parent-child relationships.
+#   We may want to preserve this information with a custom attribute.  We can also add this
+#   as a new attribute in a trace, but it will still not be seen in the UI other than a trace
+#   attribute.
+
+
+class Span(otel_api_trace.Span):
+    def __init__(
+        self,
+        name=None,
+        parent=None,  # SpanContext
+        resource=None,
+        attributes=None,
+        kind=otel_api_trace.SpanKind.INTERNAL,
+        nr_transaction=None,
+        nr_trace_type=FunctionTrace,
+        instrumenting_module=None,
+        *args,
+        **kwargs,
+    ):
+        self.name = name
+        self.otel_parent = parent
+        self.attributes = attributes or {}
+        self.kind = kind
+        self.nr_transaction = (
+            nr_transaction or current_transaction()
+        )  # This attribute is purely to prevent garbage collection
+        self.nr_trace = None
+        self.instrumenting_module = instrumenting_module
+
+        self.nr_parent = None
+        current_nr_trace = current_trace()
+        if (
+            not self.otel_parent
+            or (self.otel_parent and self.otel_parent.span_id == int(current_nr_trace.guid, 16))
+            or (self.otel_parent and isinstance(current_nr_trace, Sentinel))
+        ):
+            # Expected to come here if one of three scenarios have occured:
+            # 1. `start_as_current_span` was used.
+            # 2. `start_span` was used and the current span was explicitly set
+            #    to the newly created one.
+            # 3. Only a Sentinel Trace exists so far while still having a
+            #    remote parent. From OTel's end, this will be represented
+            #    as a `NonRecordingSpan` (and be seen as `None` at this
+            #    point). This covers cases where span is remote.
+            self.nr_parent = current_nr_trace
+        else:
+            # Not sure if there is a usecase where we could get in here
+            # but for debug purposes, we will raise an error
+            _logger.debug("Otel span and NR trace do not match nor correspond to a remote span")
+            _logger.debug("otel span: %s\nnewrelic trace: %s", self.otel_parent, current_nr_trace)
+            raise ValueError("Unexpected span parent scenario encountered")
+
+        if nr_trace_type == FunctionTrace:
```

Comment:
I have a lot of questions about these. Mainly; shouldn't we be pulling all the attributes out of the OTEL attr list and mapping the here into the appropriate NR equivalents unless there isn't a spot for them in which case they are supposed to become user attributes according to the spec?

Reply:
That part will come in a separate PR.  For now, the attributes are added as custom attributes, but the actual attribute mapping per SpanKind will come in separate PRs.  Of course, most of the time, the attributes are not added during initialization, but rather throughout the span's operation, so I will be adding logic to add these when the span is ending.